Add and distribute IQv2 information in KIP-1071

bbejeck opened the pull request:
This PR adds IQv2 information to the `StreamsGroupHeartbeat` response.

For testing, all IQv2 related integration tests have been updated (and are passing) to use the new streams protocol to exercise the newly added code.  A specific integration test and unit tests are coming.

Among the tests updated:
1. `ConsistencyVectorIntegrationTest`
2. `EosIntegrationTest` 
3. `IQv2IntegrationTest`
4. `IQv2StoreIntegrationTest`
5. `IQv2VersionedStoreIntegrationTest`
6. `LagFetchIntegrationTest` 
7. `OptimizedKTableIntegrationTest`
8. `PositionRestartIntegrationTest` 
9. `QueryableStateIntegrationTest`
10. `StoreQueryIntegrationTest`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
Fix merge conflict with First pass at adding IQv2 information

bbejeck committed:
Added IQv2 information updated IQ related integration tests

bbejeck committed:
Checkstyle fixes

bbejeck commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -110,6 +110,7 @@
 import org.apache.kafka.image.MetadataImage;
 import org.apache.kafka.image.MetadataProvenance;
 import org.apache.kafka.server.common.MetadataVersion;
+
```
checkstyle fix

bbejeck commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/taskassignor/StickyTaskAssignorTest.java:
```diff
@@ -18,6 +18,7 @@
 package org.apache.kafka.coordinator.group.taskassignor;
 
 import org.apache.kafka.coordinator.group.GroupCoordinatorConfig;
+
```
checkstyle fix

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/IQv2IntegrationTest.java:
```diff
@@ -173,7 +179,7 @@ public void afterTest() {
     @AfterAll
     public static void after() {
-        CLUSTER.stop();
+        cluster.stop();
```
Checkstyle fix this applies to all `CLUSTER` -> `cluster` updates

bbejeck commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -2378,6 +2400,32 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
+        for (Map.Entry<String, Set<Integer>> taskEntry : taskEntrySet) {
+            String subtopologyId = taskEntry.getKey();
+            List<Integer> partitions = new ArrayList<>(taskEntry.getValue());
+            ConfiguredSubtopology configuredSubtopology = group.configuredTopology().subtopologies().get(subtopologyId);
```
Went with the `ConfiguredSubtopology` since it contains any resolved regex topics.

lucasbru replied:
Yes, that is correct.

bbejeck commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -2267,7 +2269,26 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
         }
 
-        // 3. Determine the partition metadata and any internal topics if needed.
+        // Build the endpoint to topic partition information
```
Until I flush out how to only send the information to the group members that don't have it, this seems like the correct spot to build the `endpointsToPartitions`

lucasbru replied:
Yes, this is the right place. I would maybe move the code to the end of this function (since everything related to "buidling the right response" is towards the of the function). Also, consider extracting the logic into a separate method, since this one is already very long.

lucasbru commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/IQv2StoreIntegrationTest.java:
```diff
@@ -2038,7 +2043,7 @@ private static Properties streamsConfiguration(final boolean cache, final boolea
         config.put(StreamsConfig.APPLICATION_ID_CONFIG, "app-" + safeTestName);
         config.put(StreamsConfig.APPLICATION_SERVER_CONFIG, "localhost:" + (++port));
-        config.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        config.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, cluster.bootstrapServers());
```
All these integration tests don't seem to use the new protocol yet. You need to set `GROUP_PROTOCOL`.

bbejeck replied:
Thanks for the pointer I missed that one, I'll parameterize that as well.

lucasbru commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java:
```diff
@@ -78,24 +80,28 @@ public class OptimizedKTableIntegrationTest {
     public static void startCluster() throws IOException {
-        CLUSTER.start();
+        final Properties props = new Properties();
+        props.setProperty(GroupCoordinatorConfig.GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, "classic,consumer,streams");
```
How do you imagine this working on trunk? I guess we should parametrize the tests to test both the old and the new protocol?

bbejeck replied:
Yes, I'll parameterize the tests.  I took a shortcut to make sure everything worked.

bbejeck replied:
@lucasbru thinking about this more, I think it would be OK to have all the tests set the protocols accepted by the broker to all the available ones.  The parameterization should be in the Kafka Streams configuration specifying to use either the `classic` or `streams` protocol.  WDYT?

lucasbru replied:
Yes, I think you are right. Parametrizing which protocols are enabled on the cluster is overcomplicating things.

lucasbru commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -2267,7 +2269,26 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
 
-        // 3. Determine the partition metadata and any internal topics if needed.
+        // Build the endpoint to topic partition information
+        final Map<String, org.apache.kafka.coordinator.group.streams.Assignment> assignmentMap = group.targetAssignment();
```
I wonder if it would be better to use the "Current assignment" instead of the target assignment here. The current assignment will more closely model what is currently available on the node. You can find the current assignment in StreamsGroupMember.assignedActiveTasks etc. pp.

I also considered deriving the topic partitions from the reported `taskOffsets` instead of the assignment, as this is precisely the tasks that are available on the node. However, I think in the original StreamsPartititonAssignor (see `populatePartitionsByHostMaps`), we seem to just take the union of active and standby tasks. I discussed this briefly with Bruno, and it seems this optimization wouldn't really be worth it, just leaving this here in case you have a different opinion.

lucasbru commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -2378,6 +2400,32 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
+            List<Integer> partitions = new ArrayList<>(taskEntry.getValue());
+            ConfiguredSubtopology configuredSubtopology = group.configuredTopology().subtopologies().get(subtopologyId);
+            if (configuredSubtopology != null) {
+                List<StreamsGroupHeartbeatResponseData.TopicPartition> topicPartitions = Stream.concat(
```
I think there is a corner case that needs to be handled. If you look at `PartitionGrouper` in the original `StreamsPartitionAssignor` code, there may be copartitioning groups with different number of partitions within the same subtopology. So while the number of tasks is always equal to the maximal number of partitions in all source topics, there maybe a subset of the source topics that have a lower number of partition. The example would be a subtopology that merges topic A with 1 partition and topic B with 2 partitions. We will get tasks 0 and 1, but we need to make sure that the topic partitions for task 0 is {A_0,B_0} and the topic partitions for task 1 is {B_1} (because topic A does not have a second partition).

To handle this corner case, it's probably enough to remove all elements of `partitions` that exceed the number of partitions in `topic`. 

I'd suggest we add a little class for this logic, so that we can test these corner cases independently of `GroupMetadatManager`.

bbejeck replied:
Done.  The logic for calculating the `TaskPartitions` has been encapsulated in the `EndpointToPartitionsManager` class (name is definitely up for debate) and there's a unit test with it.

lucasbru reviewed (commented):
Hey Bill! This is on a good way. I left some comments, I think there are some important details that we need to take of in the code.

bbejeck committed:
Start addressing comments from initial review

bbejeck committed:
Partial implementation for adjusting topic partitions

bbejeck committed:
Added functionality to share IQ v2 information, updated integration tests to use new group orotocol.

bbejeck committed:
Intermediate state

bbejeck committed:
Get IQv2 state propogated to members and added integration test

bbejeck committed:
Refactor partition assignment logic in `EndpointToPartitionsManager`.

Simplified and modularized the logic for assigning active and standby partitions by introducing `getTopicPartitions` and utilizing `Supplier<Stream<String>>`. Updated tests to reflect these changes, including new scenarios for standby replicas and multiple configurations.

Created a new integration test for passing the IQv2 information

bbejeck committed:
Checkstyle fixes

bbejeck committed:
remove topic replication config

bbejeck committed:
Changes for adding specific standby tasks in the heartbeat response
